core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java

```diff
@@ -286,8 +285,7 @@ public synchronized void close() throws IOException {
       mBlockWorkerClientPool.clear();
 
       if (mMetricsMasterClient != null) {
-        ThreadUtils.shutdownAndAwaitTermination(mExecutorService,
-            mClientContext.getConf().getMs(PropertyKey.METRICS_CONTEXT_SHUTDOWN_TIMEOUT));
+        mMetricsHeartbeat.cancel(false);
```

Comment:
what is the purpose of this line?

Reply:
By calling this we're removing it from the scheduled context.

Imagine there are 10 clients, there will be 10 tasks scheduled in `EXECUTOR_SERVICE`. They will run based on the schedule in the configuration. When `close()` is called on one of a client object we don't want to heartbeat more. `mMetricsHeartbeat.cancel` will remove the scheduled task for that client object, leaving the `EXECUTOR_SERVICE` responsible for only the 9 remaining client objects.

Reply:
See the example in the javadoc here for a possibly better explanation of how `ScheduledFuture.cancel` works: https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ScheduledExecutorService.html